### ISSUE TITLE
RFC - Refactor (Part 1) to get all the adapters using the same API

### ISSUE THREAD

**Issue by NullVoxPopuli** [rails-api/active_model_serializers#1843]
Current Issues with adapters:

 - repeated code:
   - include logic
   - building relationships
   - building the attributes hash (from fields whitelisting, for example) 
- all the logic for retrieving fields from the cache is different per adapter
- `JSON` and `Attributes` are missing features that `JSON API` offers 
   - If adapters are what we intend them to be, they'll all be just _formatters_... sort of...

So, hopefully this can spark some new design ideas / ways to solve the above, and make AMS clean again.

Adapters should have an API modeled after JSON API 
 - `data`
  - primary serializer / resource(s)
 - `included`
   - flat list -- these objects would have the AR integration that doesn't make DB queries to belongs_to records... but in here is where the include logic would be handled, so we don't have to deal with it in the individual adapters.
 - `meta`
 - `links`
 - `errors`
 
A lot of these are already accessible to the serializer, but we've been relying on the adapter to traverse all of it.  

> All of the relationships and everything should already be traversed before getting to the adapter, so the adapter can just focus on how to format the output

An adapter would be responsible for converting each of these structures into whatever format it wishes. 

Thoughts before I start working on this? ;-)

**Comment by bf4** [comment]
? https://github.com/rails-api/active_model_serializers/blob/master/docs/rfcs/template.md

**Comment by NullVoxPopuli** [comment]
I didn't know that existed, thanks!

**Comment by bf4** [comment]
> repeated code

- serializers have an `attributes` methods. It executes `read_attribute_for_serialization` on each attribute in `self.class._attributes`.
  - the list of serialized attributes can be limited by passing in an argument `options[:fields]`
- serializers have a list of defined `associations`
  - for each associated, a serializer is found, and we call `attributes` on it.
    - we pass in to attributes `options[:includes][association_name]`, see serializer for what this does, like fields

> If adapters are what we intend them to be, they'll all be just formatters... sort of...

yup, basically

- Attributes Adapter
  - given a serializer, passes options into `attributes` and then into `associations`. Associations use a key. This is basically the behavior of an active model /active record object's `serializable_hash` method it gets from `ActiveModel::Serializers::JSON` and `ActiveModel::Serialization`.
- JSON adapter
  - is basically exactly the same as attributes adapter when `ActiveModel::Serializers::JSON.include_root_in_json` is true or when `root` is passed in. See https://github.com/rails/rails/blob/4-2-stable/activemodel/lib/active_model/serializers/json.rb#L88-L93
- JSON API adapter
  - by default `associations`, called `relationships` are not serialized unless `included`.
  - two defaults are available for attributes when no relationships are included: return just `resource_object_identifier` or full `resource_object`.
  - JSON API provides other words that other adapters may use: `filter`, `sort` and some recommendations for pagination

> all the logic for retrieving fields from the cache is different per adapter

caching should be its own layer. This is a bug. See my incomplete starts to fix this https://github.com/rails-api/active_model_serializers/compare/9477fb2%5E...bf4:remove_caching  https://github.com/rails-api/active_model_serializers/compare/9477fb2%5E...bf4:remove_caching_only_knuckles https://github.com/rails-api/active_model_serializers/compare/155033e%5E...bf4:collection_caching

**Comment by bf4** [comment]
We'll also want to discuss other concerns, not totally related, but on my mind:

- Using a serializer on it's own
- Trying to serialize a serializer
- The [default serializer like in 0.8](https://github.com/rails-api/active_model_serializers/blob/0-8-stable/lib/active_model/serializer.rb#L499-L514) and  in 0.9 the [default_serializer](https://github.com/rails-api/active_model_serializers/blob/0-9-stable/lib/active_model/default_serializer.rb), [serializable module](https://github.com/rails-api/active_model_serializers/blob/0-9-stable/lib/active_model/serializable.rb), and [serializer support](https://github.com/rails-api/active_model_serializers/blob/0-9-stable/lib/active_model/serializer_support.rb)
- Look at what options/design/words was in 0.8/0.9 and in other libs, [just like in JSON API Resources](https://github.com/rails-api/active_model_serializers/pull/1750#discussion_r64588003) and how this relates to active record sql queries
- To what extent JSON API concerns should exist in serializers, like `type` see root confusion https://github.com/rails-api/active_model_serializers/issues/1756#issuecomment-222570499 https://github.com/rails-api/active_model_serializers/pull/1794
- How this might relate to re-adding `schema` https://github.com/rails-api/active_model_serializers/issues/1430
- namespacing issues? https://github.com/rails-api/active_model_serializers/issues/1442#issuecomment-196597298

**Comment by NullVoxPopuli** [comment]
@bf4, updated with the RFC template.

**Comment by NullVoxPopuli** [comment]
So that it's recorded, I'm going to start with some PRs of refactoring some of the smaller things, such as include,fields, and see where that takes me

**Comment by bf4** [comment]
@NullVoxPopuli can you turn this into a PR?

**Comment by NullVoxPopuli** [comment]
I think so. Should I wait on your caching removal PR / any other PRs?

**Comment by bf4** [comment]
@NullVoxPopuli no, and the caching removal pr is really mean to refactor caching. I'll push the tests to it to make that more clear

**Comment by richmolj** [comment]
Spent some time looking at the JSON API adapter yesterday and here is an issue I see.

Attributes adapter has a serializer-calls-serializers pattern. One serializer [serializes itself and its relationships](https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model_serializers/adapter/attributes.rb#L7), turtles all the way down.

JSON API adapter [recurses](https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model_serializers/adapter/json_api.rb#L90) over [each relationship](https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model_serializers/adapter/json_api.rb#L239) instead. The parent serializer doesn't serialize his children, the adapter contains this logic.

This is confusing code, but actually makes a kind of sense. The Attributes adapter returns relations as a nested payload - so each serializer can just keep nesting. But JSON API sideloads everything in `included` - it's not nested data. This means there has to be a memo somewhere keeping track of all the relationships so they can eventually be output to `included`.

[Currently that memo is on the adapter](https://github.com/rails-api/active_model_serializers/blob/master/lib/active_model_serializers/adapter/json_api.rb#L252). The alternative would be sticking it on the parent serializer, but then you could make the case A) adapter (formatter) logic is leaking into the serializers, B) JSONAPI-specific logic is leaking into areas not specific to JSONAPI.

You could also follow the JSONAPI adapter's pattern in every adapter, it just means the additional complexity that comes with that pattern. You'd more or less be taking something simple, converting it to something more complex, then formatting it back as something simple again.

**Comment by bf4** [comment]
@richmolj 
> Attributes adapter has a serializer-calls-serializers pattern. One serializer serializes itself and its relationships, turtles all the way down.

> JSON API adapter recurses over each relationship instead. The parent serializer doesn't serialize his children, the adapter contains this logic.

I kind of consider this a bug in the divergence of how serializers are used.  However, to be fair-ish to history, it was my work that pushed most of the 'Attributes' code back into the serializer.
